### PR TITLE
rate limit for document uploads added

### PR DESCRIPTION
**Pull request by olusegz07**
### Checklist

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)
- [ ] Does this PR introduce a breaking change

### CODE REVIEW

File: api/documents/index.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
+// Time configuration
+const INITIAL_TIMEOUT_PERIOD = 5000; // 5 seconds
+const MAX_TIMEOUT_PERIOD = 180000; // 180 seconds
+
+let isRequestRateLimited = false;
```
Comment (andywilkinshmcts) [inline api/documents/index.ts:5]: You can't do this - it's a global variable. Needs to be a session variable

File: api/documents/index.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  return true;
+}
+
+// Handle responses being sent back to the client
+export function handleResponse(req, res, proxyRes, json) {
+  // Current time stored as the last time a document upload was completed
+  res.session.lastUploadTime = Date.now();
+
+  // Double the timeout period up to the maximum, if not rate-limited
+  if (!isRequestRateLimited) {
```
Comment (andywilkinshmcts) [inline api/documents/index.ts:37]: If we get another request while we are within the rate limit timeout, then double the timeout, up to some max value. The isRequestRateLimited value (which needs to be session specific, not global) is set to true if we get a request within the timeout period. Therefore I think this needs to be inverted - it should say:
 if (req.session.isRequestRateLimited)

File: api/documents/index.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    const nextTimeout = (res.session.nextTimeout || INITIAL_TIMEOUT_PERIOD) * 2;
+    res.session.nextTimeout = Math.min(nextTimeout, MAX_TIMEOUT_PERIOD);
+  } else {
+    delete res.session.nextTimeout;
```
Comment (andywilkinshmcts) [inline api/documents/index.ts:41]: Not sure that you want to do this - should be set to the default timeout

**Review by andywilkinshmcts — CHANGES_REQUESTED** [review]

File: api/documents/index.ts (hunk 3 above)
Comment (olusegz07) [reply api/documents/index.ts:41]: As discussed, this is valid, as code in line 15 addresses the concern raised.

  // If there is no timeout period in the session, set it to the default
  const timeoutPeriod = nextTimeout || defaultTimeoutPeriod;

**Review by OgunyemiO — APPROVED** [review]

**Review by andywilkinshmcts — DISMISSED** [review]

**Review by andywilkinshmcts — DISMISSED** [review]

**Review by OgunyemiO — DISMISSED** [review]

**Pull request merged**